jgrapht-core/src/main/java/org/jgrapht/util/DoublyLinkedList.java

```diff
@@ -17,473 +17,1387 @@
+    }
+
+    /**
+     * Returns the last {@link ListNode node} holding the specified {@code element} in this list.
+     * More formally, returns the last {@code ListNode} such that
+     * {@code Objects.equals(element, node.getValue())}, or {@code null} if there is no such node.
+     * <p>
+     * This method has linear runtime complexity O(n).
+     * </p>
+     * 
+     * @param element whose {@code ListNode} is to return
+     * @return the last {@code ListNode} holding the {@code element} or null if no node was found
+     */
+    public ListNode<E> lastNodeOf(Object element)
+    {
+        return searchNode(this::tail, n -> n.prev, element).getFirst();
+    }
+
+    /**
+     * Returns a {@link Pair} of the first encountered {@link ListNode} in this list, whose
+     * {@code value} is equal to the given {@code element}, and its index. Or if this list does not
+     * contain such node a Pair of {@code null} and {@code -1};
+     * <p>
+     * The search starts at the node supplied by {@code first} and advances in the direction induced
+     * by the specified {@code next} operator.
+     * </p>
+     * 
+     * @param first supplier of the first node to check if this list is not empty
+     * @param next {@code Function} to get from the current node the next node to check
+     * @param element for that the first node with equal value is searched.
+     * @return a {@link Pair} of the first encountered {@code ListNode} holding a {@code value}
+     *         equal to {@code element} and its index, or if no such node was found a
+     *         {@code Pair.of(null, -1)}
+     */
+    private Pair<ListNodeImpl<E>, Integer> searchNode(
+        Supplier<ListNodeImpl<E>> first, UnaryOperator<ListNodeImpl<E>> next, Object element)
+    {
+        if (!isEmpty()) {
+            int index = 0;
+            ListNodeImpl<E> firstNode = first.get();
+            ListNodeImpl<E> node = firstNode;
+            do {
+                if (Objects.equals(node.value, element)) {
+                    return Pair.of(node, index);
+                }
+                index++;
+                node = next.apply(node);
+            } while (node != firstNode);
         }
-        ++size;
-        ++modCount;
-        return result;
+        return Pair.of(null, -1);
     }
 
     /**
-     * Adds the {@code value} to the beginning of this list. Returns the list node allocated for storing
-     * the {@code value}. The returned value is the new head of the list.
-     *
-     * @param value the value to add
-     * @return the list node allocated for storing the {@code value}
+     * Inserts the specified element at the front of this list. Returns the {@link ListNode}
+     * allocated to store the {@code value}. The returned {@code ListNode} is the new head of the
+     * list.
+     * <p>
+     * This method is equivalent to {@link #addFirst(Object)} but returns the allocated
+     * {@code ListNode}.
+     * </p>
+     * 
+     * @param element to add
```

Comment:
Understand, I use Eclipse and in the way the javadoc displayed there it looks quiet OK. But I changed it.

I once tried to compile the javadoc on my machine as you said, but it failed somehow and I didn't want to spend time searching for the cause. Therefore I don't have it.

Reply:
Could you please past the error message here? It's important that users/contributors are able to compile the documentation locally.

Reply:
Sure. I run the commands in the order listed in the developer guide (Eclipse was turned of to avoid interferences):
```
mvn clean
mvn verify
mvn checkstyle:check -P checkstyle
mvn javadoc:aggregate
```
I get the following console output (I skipped repeated elements):

```
C:\Users\Hannes\development\git\repos\jgrapht>mvn javadoc:aggregate
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] JGraphT - Parent                                                   [pom]
[INFO] JGraphT - Core                                                     [jar]
[INFO] JGraphT - I/O                                                      [jar]
[INFO] JGraphT - Optimized Graphs                                         [jar]
[INFO] JGraphT - Ext                                                      [jar]
[INFO] JGraphT - Guava Adapter                                            [jar]
[INFO] JGraphT - Demo                                                     [jar]
[INFO] JGraphT - Bundle                                                   [jar]
[INFO] JGraphT - Distributable Archives                                   [pom]
[INFO]
[INFO] ------------------------< org.jgrapht:jgrapht >-------------------------
[INFO] Building JGraphT - Parent 1.3.2-SNAPSHOT                           [1/9]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] >>> maven-javadoc-plugin:3.0.0:aggregate (default-cli) > compile @ jgrapht >>>
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Parent 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht ---
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Core 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-core ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-core ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-core\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-core ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 383 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-core\target\classes
[WARNING] /C:/Users/Hannes/development/git/repos/jgrapht/jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java: Some input files use unchecked or unsafe operations.
[WARNING] /C:/Users/Hannes/development/git/repos/jgrapht/jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java: Recompile with -Xlint:unchecked for details.
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - I/O 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-io ---
[INFO]
[INFO] --- antlr4-maven-plugin:4.7:antlr4 (antlr) @ jgrapht-io ---
[INFO] No grammars to process
[INFO] ANTLR 4: Processing source directory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-io\src\main\antlr4
[INFO]
[INFO] --- replacer:1.5.3:replace (default) @ jgrapht-io ---
[INFO] Replacement run on 32 files.
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-io ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-io ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 126 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-io\target\classes
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Optimized Graphs 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-opt ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-opt ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-opt\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-opt ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 14 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-opt\target\classes
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Ext 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-ext ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-ext ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-ext\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-ext ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-ext\target\classes
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Guava Adapter 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-guava ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-guava ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-guava\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-guava ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 12 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-guava\target\classes
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Demo 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-demo ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-demo ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-demo\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-demo ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 12 source files to C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-demo\target\classes
[WARNING] /C:/Users/Hannes/development/git/repos/jgrapht/jgrapht-demo/src/main/java/org/jgrapht/demo/GraphMLDemo.java: Some input files use or override a deprecated API.
[WARNING] /C:/Users/Hannes/development/git/repos/jgrapht/jgrapht-demo/src/main/java/org/jgrapht/demo/GraphMLDemo.java: Recompile with -Xlint:deprecation for details.
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Bundle 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-bundle ---
[INFO]
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ jgrapht-bundle ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-bundle\src\main\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ jgrapht-bundle ---
[INFO] No sources to compile
[INFO]
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO] Forking JGraphT - Distributable Archives 1.3.2-SNAPSHOT
[INFO] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[INFO]
[INFO] --- maven-enforcer-plugin:1.0:enforce (enforce-maven) @ jgrapht-dist ---
[INFO]
[INFO] <<< maven-javadoc-plugin:3.0.0:aggregate (default-cli) < compile @ jgrapht <<<
[INFO]
[INFO]
[INFO] --- maven-javadoc-plugin:3.0.0:aggregate (default-cli) @ jgrapht ---
[ERROR] no module descriptor for org.jgrapht:jgrapht
[ERROR] no module descriptor for org.jgrapht:jgrapht-core
[ERROR] no module descriptor for org.jgrapht:jgrapht-io
[ERROR] no module descriptor for org.jgrapht:jgrapht-opt
[ERROR] no module descriptor for org.jgrapht:jgrapht-ext
[ERROR] no module descriptor for org.jgrapht:jgrapht-guava
[ERROR] no module descriptor for org.jgrapht:jgrapht-demo
[ERROR] no module descriptor for org.jgrapht:jgrapht-bundle
[ERROR] no module descriptor for org.jgrapht:jgrapht-dist
[INFO]
Loading source files for package org.jgrapht.alg.clique...

```
* * Repeated for each package in jGraphT * 

```
Constructing Javadoc information...
Standard Doclet version 13.0.1
Building tree for all the packages and classes...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\org\jgrapht\Graph.html...
```
* * Repeated for each file in jGraphT* 

```
Building index for all the packages and classes...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\overview-tree.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\deprecated-list.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\index.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\index-all.html...
Building index for all classes...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\allclasses-index.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\allpackages-index.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\overview-summary.html...
Generating C:\Users\Hannes\development\git\repos\jgrapht\target\site\apidocs\help-doc.html...
100 errors
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] JGraphT - Parent 1.3.2-SNAPSHOT .................... FAILURE [ 56.545 s]
[INFO] JGraphT - Core ..................................... SKIPPED
[INFO] JGraphT - I/O ...................................... SKIPPED
[INFO] JGraphT - Optimized Graphs ......................... SKIPPED
[INFO] JGraphT - Ext ...................................... SKIPPED
[INFO] JGraphT - Guava Adapter ............................ SKIPPED
[INFO] JGraphT - Demo ..................................... SKIPPED
[INFO] JGraphT - Bundle ................................... SKIPPED
[INFO] JGraphT - Distributable Archives 1.3.2-SNAPSHOT .... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 57.591 s
[INFO] Finished at: 2020-02-03T22:26:49+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.0.0:aggregate (default-cli) on project jgrapht: An error has occurred in Javadoc report generation:
[ERROR] Exit code: 1 - javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/8/docs/api/ are in the unnamed module.
[ERROR] C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-core\src\main\java\org\jgrapht\graph\AbstractGraph.java:26: error: tag not supported in the generated HTML version: tt
[ERROR]  * A skeletal implementation of the <tt>Graph</tt> interface, to minimize the effort required to
[ERROR]                                     ^
[ERROR] C:\Users\Hannes\development\git\repos\jgrapht\jgrapht-core\src\main\java\org\jgrapht\graph\AsUndirectedGraph.java:44: error: tag not supported in the generated HTML version: tt
[ERROR]  * but relies on <tt>Object</tt>'s <tt>equals</tt> and <tt>hashCode</tt> methods. This graph will be
```

Reply:
Could you please, specify, what JVM version you're using? It looks like the <tt> tag is not supported after JVM 8.

Reply:
I'm using OpenJDK13.

Reply:
All right, so the issue is that you're using a newer JVM version. I find it convenient to be able to generate the documentation from time to time in order to check what will be visible to the users. If you want to generate the documentation, you can install JVM 8 alongside your JVM 13.